src/rust/engine/graph/src/lib.rs

```diff
@@ -93,17 +94,16 @@ impl<N: Node> InnerGraph<N> {
   }
 
   fn ensure_entry(&mut self, node: N) -> EntryId {
-    InnerGraph::ensure_entry_internal(&mut self.pg, &mut self.nodes, node)
-  }
-
-  fn ensure_entry_internal(pg: &mut PGraph<N>, nodes: &mut Nodes<N>, node: N) -> EntryId {
-    if let Some(&id) = nodes.get(&node) {
+    if let Some(&id) = self.nodes.get(&node) {
+      // Update node access time
+      *(self.roots_by_age.get_mut(&node).unwrap()) = Instant::now();
```

Comment:
`ensure_entry` is called in more places than just at the root: it is also called on inner nodes. So rather than adjusting it, you'd probably only want to adjust `fn poll` and `fn create` (which are the public methods of `Graph` which allow you to request a root node) to insert into `roots_by_age`.